helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -403,16 +403,55 @@ public RoutingTableSnapshot getRoutingTableSnapshot(PropertyType propertyType, S
     return snapshots;
   }
 
+
   /**
    * Add RoutingTableChangeListener with user defined context
    * @param routingTableChangeListener
    * @param context user defined context
    */
   public void addRoutingTableChangeListener(
       final RoutingTableChangeListener routingTableChangeListener, Object context) {
+    addRoutingTableChangeListener(routingTableChangeListener, context, false);
+  }
+
+  /**
+   * Add RoutingTableChangeListener with user defined context
+   * @param routingTableChangeListener
+   * @param context user defined context
+   * @param doInit whether to trigger the initial callback during adding listener
+   */
+  public void addRoutingTableChangeListener(
+      final RoutingTableChangeListener routingTableChangeListener, Object context, boolean doInit) {
     _routingTableChangeListenerMap.put(routingTableChangeListener, new ListenerContext(context));
     logger.info("Attach RoutingTableProviderChangeListener {}",
         routingTableChangeListener.getClass().getName());
+    if (doInit) {
+      if (_sourceDataTypeMap.isEmpty()) {
```

Comment:
If `_sourceDataTypeMap ` is not properly initialized in the constructor (if this method gets called before the object is fully initialized), this might throw an NPE. I think we should add a null check here on the map. This will guard against potential concurrency issues (that would be pretty hard to debug but might happen regardless).

Reply:
I added the null check. Just to make sure, you're concerning the race condition between add routing table listener and the routing table provider construction?

Reply:
@pkuwm I think it makes sense. @narendly could you please let us know when it is possible that adding the listener happens before routing table provider is fully initialized.

Reply:
You could read this article: https://stackoverflow.com/questions/24367472/is-an-object-created-first-and-then-its-constructor-executed. Pay attention to point 4 - although taking a closer look, I don't think this leaks itself as `this`. We actually see this in the commit: https://github.com/apache/helix/commit/ff0485bdb65d1051ca63ecc97c5b864aee932465

Reply:
Thanks for the information. Didn't know this before.

Reply:
@zhangmeng916  Then are we checking null for sourceMap in validateSourceDataTypeMap?

Reply:
Sorry if my comment wasn't clear. What I was saying was since `RoutingTableProvider` doesn't give its own reference and doesn't implement `RoutingTableChangeListener`, we probably don't have to worry about this and we shouldn't have to add a null check here. I think it should be okay as is :)